### PR TITLE
NetCli: Lazily remove banned connections

### PR DESCRIPTION
**Pull request by Geod24**
```
If the connection is banned, we should just drop it and restart the loop.
```

Best viewed with whitespace diff off.

### CODE REVIEW

File: source/agora/network/Client.d [hunk 1]
```diff
@@ -604,35 +604,51 @@ public class NetworkClient
...
+            // Clients without connections should not wait and instead error out
+            if (!this.connections.length)
+                return onError();
+
+            foreach (index, conn; this.connections)
             {
                 if (this.banman.isBanned(conn.address))
                 {
-                    log.dbg("Client.attemptRequest '{}' skipped as {} is banned: {}/{}", name, conn.address, idx, this.max_retries);
+                    this.connections.remove(index);
```
Comment (linked0) [inline source/agora/network/Client.d]: We can use this connection again after the banning duration ends. So why should it be removed?

File: source/agora/network/Client.d [hunk 2]
```diff
@@ -604,35 +604,51 @@ public class NetworkClient
...
+                    log.warn("Removing banned address {} while performing {}, addresses left: {}",
+                             conn.address, name, this.addresses);
+                    continue RETRY;
```
Comment (omerfirmak) [inline source/agora/network/Client.d]: This fixes the problem only for this instance of the loop, but we are removing from a list that could potentially be traversed by multiple fibers at any given time.

Comment (omerfirmak) [reply source/agora/network/Client.d]: @Geod24 , added a new commit to fix this and rebased.

Comment (omerfirmak) [reply source/agora/network/Client.d]: we can squash if you like it.

File: source/agora/network/Client.d [hunk 3]
```diff
@@ -603,36 +606,62 @@ public class NetworkClient
...
+                    {
+                        curr_conn_version = this.connections_version;
+                        continue RETRY;
+                    }
+                }
+
                 if (this.banman.isBanned(conn.address))
                 {
-                    log.dbg("Client.attemptRequest '{}' skipped as {} is banned: {}/{}", name, conn.address, idx, this.max_retries);
+                    this.connections.remove(index);
```
Comment (omerfirmak) [inline source/agora/network/Client.d]: We shouldnt delete by index tho, it should be by value.

File: source/agora/network/Client.d [hunk 4]
```diff
@@ -624,12 +628,21 @@ public class NetworkClient
 
             foreach (index, conn; this.connections)
             {
+                scope (exit)
+                {
+                    if (this.connections_version > curr_conn_version)
+                    {
+                        curr_conn_version = this.connections_version;
+                        continue RETRY;
```
Comment (Geod24) [inline source/agora/network/Client.d]: We should not jump from a `scope (exit)`

File: source/agora/network/Client.d (hunk 3 above)
Comment (omerfirmak) [reply source/agora/network/Client.d]: done

File: source/agora/network/Client.d (hunk 4 above)
Comment (omerfirmak) [reply source/agora/network/Client.d]: updated

**Pull request merged**